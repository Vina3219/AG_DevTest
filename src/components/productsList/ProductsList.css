.productsList__container {
  display: grid;
  justify-content: center;
  align-items: center;
  background-color: #f5f5f5;
  list-style: none;
  padding: 0;
  margin: 0;
  flex: 1;
}

.productsList__container--2columns {
  grid-template-columns: 49% 49%;
}

.productsList__container--3columns {
  grid-template-columns: 32% 32% 32%;
}

.productsList__container--4columns {
  grid-template-columns: 24% 24% 24% 24%;
}

@media only screen and (max-width: 900px) {
  .productsList__container {
    grid-template-columns: 49% 49%;
  }
}

@media only screen and (max-width: 450px) {
  .productsList__container {
    grid-template-columns: 98%;
  }

  .app-footer {
    padding-right: 0;
    padding-left: 0;
    font-size: 14px;
    justify-content: center;
  }
}
