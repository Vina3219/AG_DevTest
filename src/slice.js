import { createSlice } from '@reduxjs/toolkit';
import { loadProducts, getTotal } from './services'

const initialState = {
    products: [],
    pageNum: 1,
    productsCountPerPage: 8,
    totalProductsCount: 0
}

const slice = createSlice({
    name: 'app',
    initialState,
    reducers:{
        setPageNum: (state, action) => {
            state.pageNum = action.payload.pageNum;
            state.products = loadProducts(state.pageNum, state.productsCountPerPage);
        },
        setProductsPerPage: (state, action) => {
            state.productsCountPerPage = action.payload.productsCountPerPage;
            state.pageNum = Math.ceil(state.products[0].id/action.payload.productsCountPerPage);
            state.products = loadProducts(state.pageNum, state.productsCountPerPage);
        },
        loadPage: (state) => {
            state.products = loadProducts(state.pageNum, state.productsCountPerPage);
            state.totalProductsCount = getTotal();
        }
    }
})

export const { setPageNum, setProductsPerPage: setProductsCountPerPage, loadPage } = slice.actions;

export const selectProducts = (state) => state.app.products;
export const selectPageNum = (state) => state.app.pageNum;
export const selectTotalPage = (state) => Math.ceil(state.app.totalProductsCount/state.app.productsCountPerPage);
export const selectProductsCountPerPage = (state) => state.app.productsCountPerPage;
export const selectTotalProductsCount = (state) => state.app.totalProductsCount;

export default slice.reducer;
