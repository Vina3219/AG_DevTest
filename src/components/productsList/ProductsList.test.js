import { useSelector } from "react-redux";
import { render } from "@testing-library/react";
import ProductsList from "./ProductsList";
import { selectProducts, selectProductsCountPerPage } from "../../slice";

jest.mock("react-redux", () => ({
  useSelector: jest.fn(),
}));

jest.mock("../../slice", () => ({
  ...jest.requireActual("../../slice"),
  selectProducts: jest.fn(),
  selectProductsCountPerPage: jest.fn(),
}));

jest.mock("../product/Product", () => jest.fn());

describe("ProductsList", () => {
  beforeEach(() => {
    useSelector.mockReset().mockImplementation((selector) => selector());
    selectProducts.mockReset().mockReturnValue([{ id: 1 }]);
    selectProductsCountPerPage.mockReset().mockReturnValue(8);
  });

  it("matches snapshot", () => {
    const { asFragment } = render(<ProductsList />);
    expect(asFragment()).toMatchSnapshot();
  });
});
