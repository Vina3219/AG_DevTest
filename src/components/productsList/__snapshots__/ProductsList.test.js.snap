// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsList matches snapshot 1`] = `
<DocumentFragment>
  <ol
    class="productsList__container productsList__container--4columns"
  >
    <li />
  </ol>
</DocumentFragment>
`;
